kotlin-codegen/compiler/src/main/java/com/squareup/moshi/MoshiKotlinCodeGenProcessor.kt

```diff
@@ -0,0 +1,759 @@
+              JsonQualifier::class.java)
+
+          Property(
+              name = paramName,
+              fqClassName = paramFqcn,
+              serializedName = serializedName,
+              hasDefault = valueParameter.declaresDefaultValue,
+              nullable = nullable,
+              typeName = valueParameter.type.asTypeName(nameResolver, classProto::getTypeParameter),
+              jsonQualifiers = jsonQualifiers)
+        }
+
+    val genericTypeNames = classProto.typeParameterList
+        .map {
+          val variance = it.variance.asKModifier().let {
+            // We don't redeclare out variance here
+            if (it == OUT) {
+              null
+            } else {
+              it
+            }
+          }
+          TypeVariableName.invoke(
+              name = nameResolver.getString(it.name),
+              bounds = *(it.upperBoundList
+                  .map { it.asTypeName(nameResolver, classProto::getTypeParameter) }
+                  .toTypedArray()),
+              variance = variance)
+              .reified(it.reified)
+        }.let {
+          if (it.isEmpty()) {
+            null
+          } else {
+            it
+          }
+        }
+
+    return Adapter(
+        fqClassName = fqClassName,
+        packageName = packageName,
+        propertyList = parameters,
+        originalElement = element,
+        hasCompanionObject = hasCompanionObject,
+        visibility = classProto.visibility!!,
+        genericTypeNames = genericTypeNames,
+        elementUtils = elementUtils)
+  }
+
+  private fun errorMustBeDataClass(element: Element) {
+    messager.printMessage(ERROR,
+        "@${MoshiSerializable::class.java.simpleName} can't be applied to $element: must be a Kotlin data class",
+        element)
+  }
+
+  private fun Adapter.generateAndWrite(): Boolean {
+    val adapterName = "${name}JsonAdapter"
+    val outputDir = generatedDir ?: mavenGeneratedDir(adapterName)
+    val fileBuilder = FileSpec.builder(packageName, adapterName)
+    generate(adapterName, fileBuilder)
+    fileBuilder
+        .build()
+        .writeTo(outputDir)
+    return true
+  }
+
+  private fun mavenGeneratedDir(adapterName: String): File {
+    // Hack since the maven plugin doesn't supply `kapt.kotlin.generated` option
+    // Bug filed at https://youtrack.jetbrains.com/issue/KT-22783
+    val file = filer.createSourceFile(adapterName).toUri().let(::File)
+    return file.parentFile.also { file.delete() }
+  }
+}
+
+/**
+ * Creates a joined string representation of simplified typename names.
+ */
+private fun List<TypeName>.simplifiedNames(): String {
+  return joinToString("_") { it.simplifiedName() }
+}
+
+private fun TypeName.resolveRawType(): ClassName {
+  return when (this) {
+    is ClassName -> this
+    is ParameterizedTypeName -> rawType
+    else -> throw UnsupportedOperationException("Cannot get raw type from $this")
```

Comment:
IllegalArgumentException?

Reply:
I always wondered if that still made sense in extension functions 🤔 . I guess it does!

Reply:
1fec18a